rust/lance-index/src/scalar/inverted/builder.rs

```diff
@@ -288,12 +290,26 @@ impl InnerBuilder {
     }
 
     pub async fn remap(&mut self, mapping: &HashMap<u64, Option<u64>>) -> Result<()> {
-        // no need to remap the TokenSet,
-        // no row_id is stored in the TokenSet
+        // for the docs, we need to remove the rows that are removed from the doc set,
+        // and update the row ids of the rows that are updated
         let removed = self.docs.remap(mapping);
-        for posting_list in self.posting_lists.iter_mut() {
+
+        // for the posting lists, we need to remap the doc ids:
+        // - if the a row is removed, we need to shift the doc ids of the following rows
+        // - if a row is updated (assigned a new row id), we don't need to do anything with the posting lists
+        let mut empty_posting_lists = Vec::new();
+        for (token_id, posting_list) in self.posting_lists.iter_mut().enumerate() {
             posting_list.remap(&removed);
+            if posting_list.is_empty() {
+                empty_posting_lists.push(token_id as u32);
+            }
         }
+        self.posting_lists
+            .retain(|posting_list| !posting_list.is_empty());
```

Comment:
How about we implementing in `retain` directly like:

```rust
let mut next_id: u32 = 0;
let mut empty_posting_lists = Vec::new();

self.posting_lists.retain(|pl| {
    pl.remap(&removed);

    let keep = !pl.is_empty();
    if !keep {
        empty_posting_lists.push(next_id);
    }
    next_id += 1;

    keep
});
```

In this way, we can save another `O(n)` loop.